File: algo/datastruct/tree/trie/src/Patricia.hs
type Key = String

data Patricia a = Patricia { value :: Maybe a
                           , children :: [(Key, Patricia a)]}

empty = Patrcia Nothing []

leaf :: a -> Patricia a
leaf x = Patricia (Just x) []

--insert
insert :: Patricia a -> Key -> a -> Patricia a
insert t k x = Patricia (value t) ins (children t) k x where
    ins []     k x = [(k, Patricia (Just x) [])]
    ins (p:ps) k x 
        | (fst p) == k 
            = (k, Patricia (Just x) children (snd p)):ps --overwrite
        | match (fst p) k 
            = (branch k (leaf x) (fst p) (snd p)):ps
        | otherwise 
            = p:(ins ps k x)

match :: Key -> Key -> Bool
match [] _ = False
match _ [] = False
match x y = head x == head y

branch :: Key -> Patricia a -> Key -> Patricia a -> (Key, Patricia a)
branch k1 t1 k2 t2 
    | k1'=="" = (k, Patricia (value t1) [(k2', t2)])
    | k2'=="" = (k, Patricia (value t2) [(k1', t1)])
    | otherwise = (k, Patricia Nothing [(k1', t1), (k2', t2)]) 
   where
      k = lcp k1 k2
      k1' = drop (length k) k1
      k2' = drop (length k) k2

lcp :: Key -> Key -> Key
lcp [] _ = []
lcp _ [] = []
lcp (x:xs) (y:ys) = if x==y then x:(lcp xs ys) else []

File: algo/datastruct/tree/trie/src/patricia.py
#!/usr/bin/python

import string
from trieutil import *

# Patricia Trie definition
class Patricia:
    def __init__(self, value = None):
        self.value = value
        self.children = {}

# longest common prefix
# returns (p, s1', s2'), where p is lcp, s1'=s1-p, s2'=s2-p
def lcp(s1, s2):
    j=0
    while j<=len(s1) and j<=len(s2) and s1[0:j]==s2[0:j]:
        j+=1
    j-=1
    return (s1[0:j], s1[j:], s2[j:])

def branch(key1, tree1, key2, tree2):
    if key1 == "":
        #example: insert "an" into "antoher"
        tree1.children[key2] = tree2
        return tree1
    t = Patricia()
    t.children[key1] = tree1
    t.children[key2] = tree2
    return t

def insert(t, key, value = None):
    if t is None:
        t = Patricia()

    node = t
    while(True):
        match = False
        for k, tr in node.children.items():
            if key == k: # just overwrite
                node.value = value
                return t
            (prefix, k1, k2)=lcp(key, k)
            if prefix != "":
                match = True
                if k2 == "": 
                    # example: insert "antoher" into "an", go on traversing
                    node = tr
                    key = k1
                    break
                else: #branch out a new leaf
                    node.children[prefix] = branch(k1, Patricia(value), k2, tr)
                    del node.children[k]
                    return t
        if not match: # add a new leaf
            node.children[key] = Patricia(value)
            break
    return t

def lookup(t, key):
    if t is None:
        return None
    while(True):
        match = False
        for k, tr in t.children.items():
            if k == key:
                return tr.value
            (prefix, k1, k2) = lcp(key, k)
            if prefix != "" and k2 == "":
                match = True
                key = k1
                t = tr
                break
        if not match:
            return None

def to_string(t):
    return trie_to_str(t)

def list_to_patricia(l):
    return from_list(l, insert)

def map_to_patricia(m):
    return from_map(m, insert)

class PatriciaTest:
    def run(self):
        self.test_lcp()
        self.test_insert()
        self.test_lookup()

    def test_lcp(self):
        print "test lcp"
        print lcp("om", "ub")
        print lcp("romane", "romanus")
        print lcp("an", "another")

    def test_insert(self):
        print "test insert"
        t = list_to_patricia(["a", "an", "another", "b", "bob", "bool", "home"])
        print to_string(t)
        t = list_to_patricia(["romane", "romanus", "romulus"])
        print to_string(t)
        t = map_to_patricia({"001":'y', "100":'x', "101":'z'})
        print to_string(t)
        t = list_to_patricia(["home", "bool", "bob", "b", "another", "an", "a"]);
        print to_string(t)

    def test_lookup(self):
        print "test lookup"
        t = map_to_patricia({"a":1, "an":2, "another":7, "boy":3, "bool":4, "home":4})
        print "search t another", lookup(t, "another")
        print "search t boo", lookup(t, "boo")
        print "search t boy", lookup(t, "boy")
        print "search t by", lookup(t, "by")
        print "search t boolean", lookup(t, "boolean")

if __name__ == "__main__":
    PatriciaTest().run()
    
